Streamline ParseType Logic

pi1024e opened the pull request:
<!-- What's in this pull request? -->
The ParseType code has a lot of jumps and incongruity in respect to the code's behavior despite there being functions that have identical behavior. For readability, and to highlight this, I refactored this file to better reflect that.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

slavapestov commented:
@swift-ci Please smoke test

pi1024e commented:
Reran the tests and I realized I made a typo. Now this compiles and works.

rintaro commented on lib/Parse/ParseType.cpp:
```diff
@@ -1375,29 +1373,29 @@ bool Parser::isOptionalToken(const Token &T) const {
-    return true;
-  return false;
+
+  if (!T.is(tok::oper_postfix) && !T.is(tok::oper_binary_unspaced)) {
+    return false;
+  }
+  // We'll munch off the '!', so long as it is left-bound with
+  // the type (i.e., parsed as a postfix or unspaced binary operator).
+  return T.getText().startswith("!");
```
Could you keep the first half of the comment?
```
A postfix or bound infix operator token that begins with '!' can be
implicitly unwrapped optional too.
```

rintaro commented on lib/Parse/ParseType.cpp:
```diff
@@ -1682,13 +1677,9 @@ bool Parser::canParseTypeIdentifier() {
   // Qualified name base types must be followed by a period.
   // If the next token starts with a period, return true.
-  return startsWithSymbol(Tok, '.');
+  return (canParseSimpleTypeIdentifier() && startsWithSymbol(Tok, '.'));
```
I prefer the original. `if (!canParseSimpleTypeIdentifier()) return false;` is more like an early return than combined condition.

rintaro reviewed (commented):
Thank you for the improvement! Just a few requests.

pi1024e commented:
Done!

pi1024e commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -625,27 +621,32 @@ ParserResult<TypeRepr> Parser::parseDeclResultType(Diag<> MessageID) {
+      consumeToken();
+      return makeParserErrorResult(new (Context)
+                                       ErrorTypeRepr(getTypeErrorLoc()));
+    } else if (Tok.is(tok::colon)) {
```
`else` is not needed here, right?

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -625,27 +621,32 @@ ParserResult<TypeRepr> Parser::parseDeclResultType(Diag<> MessageID) {
+
+      auto diag = diagnose(colonTok, diag::extra_colon);
+      diag.fixItInsert(result.get()->getStartLoc(),
+                       getTokenText(tok::l_square));
```
Why did you wrap here? It seems like the original should fit all in one line, no?

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -625,27 +621,32 @@ ParserResult<TypeRepr> Parser::parseDeclResultType(Diag<> MessageID) {
+    if (Tok.is(tok::r_square)) {
+      auto diag = diagnose(Tok, diag::extra_rbracket);
+      diag.fixItInsert(result.get()->getStartLoc(),
+                       getTokenText(tok::l_square));
```
Same question here--why did you wrap the existing line?

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -775,28 +776,26 @@ Parser::parseTypeIdentifier(bool isParsingQualifiedDeclBaseType) {
-        continue;
+      if (peekToken().isContextualKeyword("Type") ||
+          peekToken().isContextualKeyword("Protocol"))
+        break;
```
This doesn't seem right. In the old code, we would proceed to evaluate `Tok.is(tok::code_complete)` here, not break. I don't think your change preserves the behavior of the code, or if it does, it's not obvious to me.

pi1024e replied:
This is the same exact behavior. It is just that the break was not obvious in the old code at all, which is why this change may seem confusing, but this has the exact same behavior

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -1375,14 +1374,15 @@ bool Parser::isOptionalToken(const Token &T) const {
+    return false;
+  }
+
   // A postfix or bound infix operator token that begins with '?' can be
```
I think it's clearer if you restore this to the way it was, since the comment applies to the code moved above. While the logic may be the same (I'd have to think about it), it's literally breaking up one test into two separate thoughts.

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -1391,13 +1391,13 @@ bool Parser::isImplicitlyUnwrappedOptionalToken(const Token &T) const {
+  }
+  // We'll munch off the '!', so long as it is left-bound with
+  // the type (i.e., parsed as a postfix or unspaced binary operator).
+  return T.getText().startswith("!");
```
Ditto here.

xwu reviewed (dismissed):
Hi @pi1024e, while it's admirable that you're persisting in working on some of these clean-ups, as a matter of etiquette, it's not particularly appropriate to abandon PRs with existing review comments unaddressed (#33231, #32545) and then reopen a new PR. This is the second occurrence. Just as it's nice that you're taking the time to make these changes, others are taking the time to help you improve on them.

I've reiterated a few points that came to mind on previous PRs. Please address any others (including from other reviewers too) that are not yet reflected in your changes here.

pi1024e commented:
The last PR wasn't closed by me, which is why I opened this new one.

pi1024e commented:
@xwu Done!

pi1024e commented:
@swift-ci please smoke test

xwu reviewed (commented):
Thanks for addressing comments! This is an improvement, I think.

xwu commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@pi1024e Please test it locally before pushing the changes 🙏

gottesmm committed:
[ownership] Now that we allow reborrows eliminate isForwardingSubValue it is dead.

[cxx-interop] Support reference types in function templates. (#34536)

[cxx-interop] Don't crash for unkown inline-operator. (#34602)

Skip unkown or unimportable inline operators (such as deleted operator members) instead of crashing.

Streamline ParseType Logic

xedin committed:
[ResultBuilders] Account of a fact that re-write of a statement can fail

If one of the statements in the result builder body fails to
apply solution, let's fail entire rewrite attempt, otherwise
type-checker would end up with AST that has null pointers for
some child nodes.

Resolves: rdar://problem/70256351

kavon committed:
add the -disable-implicit-concurrency-module-import hidden flag

kavon committed:
when building the swift libs, do not automatically import _Concurrency

This fixes issues when building Swift with concurrency enabled by default. See rdar://71045297

If some of our overlays need to use the concurrency API, they can manually import _Concurrency.

kavon committed:
add regression test for -disable-implicit-concurrency-module-import

pi1024e commented:
@rintaro done

pi1024e commented:
Closing because I messed up the git tree
